### PR TITLE
Fix deadlock in weavedns

### LINKED ISSUE #923: 240_dns_add_name_test.sh timing out since merge of #917
Failure since merge of #917

[first failure](https://circleci.com/gh/weaveworks/weave/927) - it's been failing ever since.

### PR DESCRIPTION
**Pull request by bboreham**
Release lock before sending on a channel.
Fixes #923 and #924.

### CODE REVIEW

**Comment by inercia** [comment]
LGTM

File: nameserver/zone_lookup.go [hunk 1]
```diff
@@ -235,12 +235,11 @@ func (zone *ZoneDb) DomainLookupInaddr(inaddr string) (res []ZoneRecord, err err
...
 		zone.mx.Lock()
-		defer zone.mx.Unlock()
-
 		// check if we should enqueue a refresh request for this name
 		n := zone.getNameSet(defaultRemoteIdent).getName(name, true)
+		now := zone.clock.Now()
+		zone.mx.Unlock() // Don't hold the lock while talking to the SchedQueue
 		if n.lastRefreshTime.IsZero() {
-			now := zone.clock.Now()
 			n.lastRefreshTime = now
```
Comment (paulbellamy) [inline nameserver/zone_lookup.go]: This is now outside the mutex, so could race.

Comment (inercia) [inline nameserver/zone_lookup.go]: Yes, the name could have been removed since we got it with `getName()`...

Comment (inercia) [inline nameserver/zone_lookup.go]: Setting the lock in a different way, all tests pass in my machine...

```Go
// Check if we must start updating a name and, in that case, trigger a immediate update
func (zone *ZoneDb) startUpdatingName(name string) {
	if zone.refreshInterval > 0 {
		zone.mx.Lock()

		// check if we should enqueue a refresh request for this name
		n := zone.getNameSet(defaultRemoteIdent).getName(name, true)
		if n.lastRefreshTime.IsZero() {
			now := zone.clock.Now()
			n.lastRefreshTime = now
			zone.mx.Unlock()

			Debug.Printf("[zonedb] Creating new immediate refresh request for '%s'", name)
			zone.refreshScheds.Add(func() time.Time { return zone.updater(name) }, now)
		} else {
			zone.mx.Unlock()
		}
	}
}
```

**Pull request merged**